app/src/main/java/com/yalta/services/LoginService.kt

```diff
@@ -1,18 +1,23 @@
 package com.yalta.services
 
 import com.yalta.repositories.*
+import java.util.*
 
 class LoginService(private val repo: LoginRepo) {
-    suspend fun login(login: String, password: String): Boolean {
+    suspend fun login(login: String, password: String): Optional<Boolean> {
         return process(
             { repo.login(login, password) },
             { sLogin ->
                 SessionService.setSession(sLogin.token)
-                true
+                Optional.of(true)
             },
             {
                 SessionService.discardSession()
-                false
+                if (it.reason == Reason.BAD_CODE) {
+                    Optional.of(false)
+                } else {
+                    Optional.empty()
+                }
```

Comment:
So this should be copy-pasted into all services? I don't really like the idea...

Reply:
Don't see any reasons to add it to all of them
For now there are only two cases where we want to differentiate failure reasons, so ok for me